pkgs/by-name/ba/bambu-unwrapped/package.nix

```diff
@@ -0,0 +1,237 @@
+assert lib.assertMsg (builtins.length gccFrontendsFlags != 0) "There needs to be at least one GCC frontend specified";
+
+# Assert that the passed clang and GCC frontends are i686
+assert builtins.all
+  (llvmPackages:
+  lib.assertMsg (llvmPackages.clang.stdenv.hostPlatform.system == "i686-linux") "Bambu frontends have to be i686 but ${llvmPackages.clang.name} is ${llvmPackages.clang.stdenv.hostPlatform.system}")
+  clangFrontends;
+assert builtins.all
+  (gcc:
+  lib.assertMsg (gcc.stdenv.hostPlatform.system == "i686-linux") "Bambu frontends have to be i686 but ${gcc.name} is ${gcc.stdenv.hostPlatform.system}")
+  gccFrontends;
+
+# Assert that the clang and GCC frontends versions are compatible with bambu
+assert builtins.all
+  (llvmPackages:
+  lib.assertMsg (builtins.elem (lib.versions.major llvmPackages.clang.version) supportedLlvmMajorVersions) "Clang version ${llvmPackages.clang.version} is not supported as a frontend by bambu. The following major clang versions are supported: ${builtins.concatStringsSep " " supportedLlvmMajorVersions}")
+  clangFrontends;
+assert builtins.all
+  (gcc:
+  lib.assertMsg (builtins.elem (lib.versions.major gcc.version) supportedGccMajorVersions) "GCC version ${gcc.version} is not supported as a frontend by bambu. The following major GCC versions are supported: ${builtins.concatStringsSep " " supportedGccMajorVersions}")
+  gccFrontends;
+
+# Assert that there is only one compiler frontend for each major version of the compiler
+let
+  passedClangVersions = (builtins.map (llvmPackages: llvmPackages.clang.version) clangFrontends);
+  duplicateClangVersion = lib.lists.findFirst (clangVersion: (builtins.length (builtins.filter (otherClangVersion: (lib.versions.major clangVersion) == (lib.versions.major otherClangVersion)) passedClangVersions)) > 1) false passedClangVersions;
+  passedGccVersions = (builtins.map (gcc: gcc.version) gccFrontends);
+  duplicateGccVersion = lib.lists.findFirst (gccVersion: (builtins.length (builtins.filter (otherGccVersion: (lib.versions.major gccVersion) == (lib.versions.major otherGccVersion)) passedGccVersions)) > 1) false passedGccVersions;
+in
+assert lib.assertMsg (duplicateClangVersion == false) "Multiple clang frontends with the same major version (${lib.versions.major duplicateClangVersion}) are not supported by bambu. You passed packages with the following versions: ${builtins.concatStringsSep " " (builtins.filter (clangVersion: (lib.versions.major clangVersion) == (lib.versions.major duplicateClangVersion)) passedClangVersions)}";
+
+assert lib.assertMsg (duplicateGccVersion == false) "Multiple gcc frontends with the same major version (${lib.versions.major duplicateGccVersion}) are not supported by bambu. You passed packages with the following versions: ${builtins.concatStringsSep " " (builtins.filter (gccVersion: (lib.versions.major gccVersion) == (lib.versions.major duplicateGccVersion)) passedGccVersions)}";
```

Comment:
```suggestion
  duplicateClangVersion = lib.lists.findFirst (clangVersion: (builtins.length (builtins.filter (otherClangVersion: (lib.versions.major clangVersion) == (lib.versions.major otherClangVersion)) passedClangVersions)) > 1) null passedClangVersions;
  passedGccVersions = (builtins.map (gcc: gcc.version) gccFrontends);
  duplicateGccVersion = lib.lists.findFirst (gccVersion: (builtins.length (builtins.filter (otherGccVersion: (lib.versions.major gccVersion) == (lib.versions.major otherGccVersion)) passedGccVersions)) > 1) null passedGccVersions;
in
assert lib.assertMsg (duplicateClangVersion == null) "Multiple clang frontends with the same major version (${lib.versions.major duplicateClangVersion}) are not supported by bambu. You passed packages with the following versions: ${builtins.concatStringsSep " " (builtins.filter (clangVersion: (lib.versions.major clangVersion) == (lib.versions.major duplicateClangVersion)) passedClangVersions)}";

assert lib.assertMsg (duplicateGccVersion == null) "Multiple gcc frontends with the same major version (${lib.versions.major duplicateGccVersion}) are not supported by bambu. You passed packages with the following versions: ${builtins.concatStringsSep " " (builtins.filter (gccVersion: (lib.versions.major gccVersion) == (lib.versions.major duplicateGccVersion)) passedGccVersions)}";
```

`statix check` gives a `[W01] Warning: Unnecessary comparison with boolean` 

`statix fix` would change `(duplicateClangVersion == false)` to `(!duplicateClangVersion)` which will break with `error: value is a string while a Boolean was expected`

Reply:
The code seems a lot cleaner with null instead of false. Until now I have only used nixpkgs-hammering for checking, do you know of any other good static analysis tools for nix?

Reply:
https://discourse.nixos.org/t/list-of-nix-linters/19279